File: Python/vdASCIIFile.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
@author: Florian Timm
@version: 2017.11.27
"""

from vdFile import VdFile
from vdPoint import VdPoint

class VdASCIIFile(VdFile):
    """ abstract class for writing ascii files """

    def _open_ascii(self, filename, file_format):
        """
        opens ascii file for writing
        :param filename: file name
        :type filename: str
        """
        filename = self._make_filename(file_format, filename)
        self.__file = open(filename, 'a')

    def _write2file(self, data):
        """
        writes ascii data to file
        :param data: data to write
        :type data: str
        """
        self.__file.write(data)

    def write(self):
        """ writes data to file """
        txt = ""
        for d in self.writing_queue:
            txt += self._format(d)
        self._write2file(txt)
        self.clear_writing_queue()

    @abstractmethod
    def _format(self, p):
        """
        abstract method: should convert point data to string for ascii file
        :param p: Point
        :type p: VdPoint
        :return: point data as string
        :rtype: str
        """
        pass

    def close(self):
        """ close file """
        self.__file.close()

File: C++/src/VdASCIIFile.cpp
/*!
 *  \brief		class for writing data to ascii file
 *
 *  \author		Florian Timm
 *  \version	2017.11.29
 *  \copyright	MIT License
 */

#include <string>
#include <stdio.h>
#include <vector>
#include <fstream>
#include <iostream>
using namespace std;
#include "VdPoint.h"
#include "VdASCIIFile.h"
#include "iniparser/iniparser.h"

void VdASCIIFile::openASCII(string filename, string file_format) {
	/**
	 * opens ascii file for writing
	 * @param filename file name
	 */
	filename = this->makeFilename(file_format, filename);
	this->file = fopen(filename.c_str(), "at");
}

void VdASCIIFile::write2file(string data) {
	/**
	 * writes ascii data to file
	 * @param data data to write
	 */
	if (this->file != NULL) {
		fputs(data.c_str(), this->file);
	}
}

void VdASCIIFile::write() {
	/** writes data to file */
	string txt;
	for (VdPoint p : this->writingQueue)
		txt += this->format(p);
	this->write2file(txt);
	this->clearWritingQueue();
}

void VdASCIIFile::close() {
	/** close file */
	fclose(this->file);
}
